File: infullmvp-basetest/src/main/java/com/infullmobile/android/infullmvp/basetest/InFullMvpFragmentBaseTest.java
package com.infullmobile.android.infullmvp.basetest;

import android.support.annotation.NonNull;

import com.infullmobile.android.infullmvp.InFullMvpFragment;

import org.junit.Before;
import org.robolectric.shadows.support.v4.SupportFragmentController;

public abstract class InFullMvpFragmentBaseTest<T extends InFullMvpFragment> {

    private T fragment;

    @Before
    public void setUp() {
        SupportFragmentController<T> fragmentController
                = SupportFragmentController.of(provideFragment());
        fragment = fragmentController.create().get();
        substituteModules(fragment);
        fragmentController.start().resume().visible();
    }

    @NonNull
    protected abstract T provideFragment();


    public T getTestedFragment() {
        return fragment;
    }

    public void substituteModules(T fragment) {
        /* NO OP */
    }
}

File: infullmvp-kotlin-basetest/src/main/java/com/infullmobile/android/infullmvp/basetest/InFullMvpFragmentBaseTest.kt
package com.infullmobile.android.infullmvp.basetest

import com.infullmobile.android.infullmvp.InFullMvpFragment
import com.infullmobile.android.infullmvp.PresentedFragmentView
import com.infullmobile.android.infullmvp.Presenter

import org.junit.Before
import org.robolectric.RuntimeEnvironment
import org.robolectric.shadows.support.v4.SupportFragmentController

abstract class InFullMvpFragmentBaseTest<T : InFullMvpFragment<*, *>> {

    lateinit var testedFragment: T
        private set

    val testedPresenter: Presenter<*>
        get() = testedFragment.presenter
    val testedView: PresentedFragmentView<*>
        get() = testedFragment.presentedView

    @Before
    open fun setUp() {
        val fragmentController = SupportFragmentController.of(provideFragment())
        testedFragment = fragmentController.create().get()
        substituteModules(testedFragment)
        fragmentController.start().resume().visible()
    }

    protected fun getString(stringResourceId: Int): String {
        return RuntimeEnvironment.application.resources.getString(stringResourceId)
    }

    protected abstract fun provideFragment(): T

    open fun substituteModules(fragment: T) {
        /* NO OP */
    }
}
